### PR TITLE
Bpapi gateway

### PR DESCRIPTION
**Pull request by k32**

### CODE REVIEW

File: apps/emqx_gateway/src/emqx_gateway_cm.erl [hunk 1]
```diff
@@ -186,36 +207,43 @@ get_chan_stats(GwName, ClientId) ->
...
+        error:badarg -> false
+    end.
+
 -spec set_chan_stats(gateway_name(),
                      emqx_types:clientid(),
                      pid(),
                      emqx_types:stats()) -> boolean().
-set_chan_stats(GwName, ClientId, ChanPid, Stats)
-  when node(ChanPid) == node() ->
-    Chan = {ClientId, self()},
```
Comment (k32) [inline apps/emqx_gateway/src/emqx_gateway_cm.erl:212]: I am pretty sure it was a bug, since this function runs in the context of the RPC dispatcher process when called remotely.

File: apps/emqx_gateway/src/proto/emqx_gateway_cm_proto_v1.erl [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+    rpc:call(node(ChanPid), emqx_gateway_cm, do_get_chan_stats, [GwName, ClientId, ChanPid]).
+
+-spec set_chan_stats(emqx_gateway_cm:gateway_name(),
+                     emqx_types:clientid(),
+                     pid(),
+                     emqx_types:stats()) -> boolean() | {badrpc, _}.
+set_chan_stats(GwName, ClientId, ChanPid, Stats) ->
+    rpc:call(node(ChanPid), emqx_gateway_cm, do_set_chan_stats, [GwName, ClientId, ChanPid, Stats]).
+
+-spec discard_session(emqx_gateway_cm:gateway_name(), emqx_types:clientid(), pid()) -> _.
```
Comment (k32) [inline apps/emqx_gateway/src/proto/emqx_gateway_cm_proto_v1.erl:61]: It's hard to guess the return type here, since the target calls an unspecified module.

**Review by thalesmg — APPROVED** [review]

File: apps/emqx_gateway/src/emqx_gateway_http.erl [hunk 3]
```diff
@@ -231,41 +229,26 @@ confexp({error, Reason}) -> error(Reason).
...
 -spec kickout_client(gateway_name(), emqx_types:clientid())
     -> {error, any()}
      | ok.
 kickout_client(GwName, ClientId) ->
-    Results = [kickout_client(Node, GwName, ClientId)
-               || Node <- mria_mnesia:running_nodes()],
-    case lists:any(fun(Item) -> Item =:= ok end, Results) of
+    Results = [emqx_gateway_cm:kick_session(GwName, ClientId, Pid)
+               || Pid <- emqx_gateway_cm:lookup_by_clientid(GwName, ClientId)],
+    case Results =:= [] orelse lists:any(fun(Item) -> Item =:= ok end, Results) of
```
Comment (thalesmg) [inline apps/emqx_gateway/src/emqx_gateway_http.erl:247]: Just curiosity: would it be simple to add a test case for empty results?  I understand that this is just a refactor and it might get too big a PR, so it might not be good to do it here.

**Review by thalesmg — APPROVED** [review]

File: apps/emqx_gateway/src/emqx_gateway_cm.erl [hunk 4]
```diff
@@ -146,16 +161,34 @@ get_chan_info(GwName, ClientId) ->
...
 get_chan_info(GwName, ClientId, ChanPid) ->
-    rpc_call(node(ChanPid), get_chan_info, [GwName, ClientId, ChanPid]).
+    wrap_rpc(emqx_gateway_cm_proto_v1:get_chan_info(GwName, ClientId, ChanPid)).
+
+-spec lookup_by_clientid(gateway_name(), emqx_types:clientid()) ->
+          [pid()].
+lookup_by_clientid(GwName, ClientId) ->
+    Nodes = mria_mnesia:running_nodes(),
+    Pids = emqx_gateway_cm_proto_v1:lookup_by_clientid(Nodes, GwName, ClientId),
+    lists:append([Pid || {ok, Pid} <- Pids]).
```
[the remark this answers is not in the document]
Comment (k32) [reply apps/emqx_gateway/src/emqx_gateway_cm.erl:191]: Yes, I was looking at the old behavior, which could crash on `badrpc`. Ignoring badrpc seems like this seems a more reasoable behavior.

[the remark this answers is not in the document]
Comment (DDDHuang) [reply apps/emqx_gateway/src/emqx_gateway_cm.erl:191]: In the API design, the current rule is that if the caller specifies a remote node, then badrpc should first check if the node name is in the cluster. If the caller does not specify a remote node or wants to operate in the whole cluster, and there are corrupted nodes in the cluster, API corruption is unavoidable. It should expose that the cluster is not working properly, not cover up the failed node error

**Review by HJianBo — DISMISSED** [review]

**Review by thalesmg — APPROVED** [review]
